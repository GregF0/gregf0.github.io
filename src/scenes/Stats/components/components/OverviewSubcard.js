import React, { useState, useEffect } from 'react'
import { useSelector } from 'react-redux'
import styled, { css } from 'styled-components'
import {
  CardContent as MatCardContent,
  Grid,
  Box,
  Card as MatCard,
  useMediaQuery,
  useTheme,
} from '@material-ui/core'
import { TrendIcon } from './components/TrendIcon'

export const OverviewSubcard = ({ icon, sum, label, trend = 0 }) => {
  const theme = useTheme()
  const isMediumScreen = +useMediaQuery(theme.breakpoints.up('md'))
  const darkMode = useSelector((state) => +state.settings.darkMode)

  const [trendColor, setTrendColor] = useState('rgba(0,0,0,0)')
  const [trendIconPath, setTrendIconPath] = useState()

  useEffect(() => {
    if (trend === 0) return
    let color, path

    if (trend > 0) {
      if (darkMode) {
        color = trendProps.positive.dark
      } else {
        color = trendProps.positive.light
      }
      path = trendProps.positive.path
    } else {
      if (darkMode) {
        color = trendProps.negative.dark
      } else {
        color = trendProps.negative.light
      }
      path = trendProps.negative.path
    }

    setTrendColor(color)
    setTrendIconPath(path)
  }, [darkMode, trend])

  return (
    <DataCard md={isMediumScreen} dark={darkMode}>
      <CardContent md={isMediumScreen}>
        <MainGrid
          container
          direction={isMediumScreen ? 'column' : 'row'}
          spacing={2}
          alignItems="center"
        >
          <Grid item xs={2} md={5} lg={9}>
            <Icon
              max-width="100%"
              src={icon}
              alt="Today Icon"
              md={isMediumScreen}
            />
          </Grid>
          <Grid item xs={10} md={12}>
            <Sum dark={darkMode} md={isMediumScreen}>
              {sum}
            </Sum>
            <Label color={theme.palette.text.secondary} md={isMediumScreen}>
              {label}
            </Label>

            {trend !== 0 && (
              <Box
                display="flex"
                alignItems="center"
                justifyContent={isMediumScreen ? 'center' : 'flex-start'}
              >
                <TrendIcon fill={trendColor} path={trendIconPath} />
                <Trend md={isMediumScreen} color={trendColor}>
                  {Math.abs(trend)}%
                </Trend>
              </Box>
            )}
          </Grid>
        </MainGrid>
      </CardContent>
    </DataCard>
  )
}

const DataCard = styled(MatCard)`
  box-shadow: 0px 3px 3px -2px rgba(0, 0, 0, 0.2),
    0px 3px 4px 0px rgba(0, 0, 0, 0.14), 0px 1px 8px 0px rgba(0, 0, 0, 0.12);
  background: inherit;

  ${({ md }) =>
    md
      ? css`
          width: 200px;
          height: 200px;
          margin-bottom: 20px;
          background: ${({ dark }) => (dark ? '#2c2c2c' : 'inherit')};
        `
      : css`
          box-shadow: none;
          margin: 0 16px;
        `}
`

const CardContent = styled(MatCardContent)`
  ${({ md }) =>
    !md &&
    css`
      padding: 8px 0 !important;
    `}
`

const MainGrid = styled(Grid)``

const Icon = styled.img`
  width: 100%;
  max-width: ${({ md }) => (md ? '50px' : '40px')};
`

const Sum = styled.span`
  display: block;
  font-weight: bold;
  font-size: 1rem;
  color: ${({ dark }) => (dark ? '#e4e4e4' : '#4e4e4e')};

  ${({ md }) =>
    md &&
    css`
      font-size: 2rem;
      text-align: center;
    `}
`

const Label = styled.span`
  font-size: 0.8rem;
  color: ${({ color }) => color};
  display: block;
  margin-bottom: 5px;

  ${({ md }) =>
    md &&
    css`
      text-align: center;
      font-size: 0.9rem;
    `}
`

const Trend = styled.span`
  display: block;
  font-weight: bold;
  font-size: 0.9rem;
  color: ${({ color }) => color};
  margin-left: 5px;

  ${({ md }) =>
    md &&
    css`
      margin-left: 7px;
      font-size: 1rem;
    `}
`

const trendProps = {
  positive: {
    light: '#3BC83C',
    dark: '#A5D6A7',
    path: 'M6 0.806397L12 6.8064L0 6.8064L6 0.806397Z',
  },
  negative: {
    light: '#F12524',
    dark: '#EF9A9A',
    path: 'M6 6.8064L0 0.806396L12 0.806399L6 6.8064Z',
  },
}

export { Sum }
