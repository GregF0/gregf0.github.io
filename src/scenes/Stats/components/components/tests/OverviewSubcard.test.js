import React from 'react'
import * as redux from 'react-redux'
import { createShallow } from '@material-ui/core/test-utils'
import { OverviewSubcard } from '../OverviewSubcard'
import icon from '../../assets/icons/week/week-light.svg'

describe('<OverviewSubcard />', () => {
  const shallow = createShallow()
  const createWrapper = () => {
    return shallow(
      <OverviewSubcard icon={icon} sum="13h 5m" label="Week" trend="9" />
    )
  }

  const createStore = () => {
    const store = {
      settings: { darkMode: false },
    }

    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((callback) => callback(store))

    return store
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('should render <OverviewSubcard /> correctly', () => {
    createStore()
    expect(createWrapper()).toMatchSnapshot()
  })
})
