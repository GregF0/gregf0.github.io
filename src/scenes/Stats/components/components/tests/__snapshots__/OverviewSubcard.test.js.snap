// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OverviewSubcard /> should render <OverviewSubcard /> correctly 1`] = `
<Styled(WithStyles(ForwardRef(Card)))
  dark={0}
  md={0}
>
  <Styled(WithStyles(ForwardRef(CardContent)))
    md={0}
  >
    <Styled(WithStyles(ForwardRef(Grid)))
      alignItems="center"
      container={true}
      direction="row"
      spacing={2}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        lg={9}
        md={5}
        xs={2}
      >
        <styled.img
          alt="Today Icon"
          max-width="100%"
          md={0}
          src="week-light.svg"
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        md={12}
        xs={10}
      >
        <styled.span
          dark={0}
          md={0}
        >
          13h 5m
        </styled.span>
        <styled.span
          color="rgba(0, 0, 0, 0.54)"
          md={0}
        >
          Week
        </styled.span>
        <Styled(MuiBox)
          alignItems="center"
          display="flex"
          justifyContent="flex-start"
        >
          <TrendIcon
            fill="rgba(0,0,0,0)"
          />
          <styled.span
            color="rgba(0,0,0,0)"
            md={0}
          >
            9
            %
          </styled.span>
        </Styled(MuiBox)>
      </WithStyles(ForwardRef(Grid))>
    </Styled(WithStyles(ForwardRef(Grid)))>
  </Styled(WithStyles(ForwardRef(CardContent)))>
</Styled(WithStyles(ForwardRef(Card)))>
`;
